src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -1332,25 +1327,52 @@ void uct_dc_mlx5_ep_handle_failure(uct_dc_mlx5_ep_t *ep, void *arg,
         ucs_debug("got error on DC flow-control endpoint, iface %p: %s", iface,
                   ucs_status_string(ep_status));
     } else {
-        status  = ib_iface->ops->set_ep_failed(ib_iface, &ep->super.super,
-                                               ep_status);
-        log_lvl = uct_ib_iface_failure_log_level(ib_iface, status, ep_status);
+        uct_dc_mlx5_iface_set_ep_failed(iface, ep, (struct mlx5_cqe64*)arg,
+                                        txwq, ep_status);
+    }
+}
 
-        if (ep_status != UCS_ERR_CANCELED) {
-            uct_ib_mlx5_completion_with_err(ib_iface, arg,
-                                            &iface->tx.dcis[dci].txwq, log_lvl);
-        }
+ucs_status_t
+uct_dc_mlx5_ep_check(uct_ep_h tl_ep, unsigned flags, uct_completion_t *comp)
+{
+    uct_dc_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_dc_mlx5_iface_t);
+    uct_dc_mlx5_ep_t *ep       = ucs_derived_of(tl_ep, uct_dc_mlx5_ep_t);
+    uint64_t dummy             = 0;
+    ucs_status_t status;
+    uct_rc_iface_send_op_t *op;
+    UCT_DC_MLX5_TXQP_DECL(txqp, txwq);
+
+    UCT_CHECK_PARAM(comp == NULL, "Unsupported completion on ep_check");
+    UCT_CHECK_PARAM(flags == 0, "Unsupported flags: %x", flags);
+
+    if ((ep->dci != UCT_DC_MLX5_EP_NO_DCI) &&
+        uct_dc_mlx5_iface_dci_has_outstanding(iface, ep->dci)) {
```

Comment:
maybe just check ep->dci != UCT_DC_MLX5_EP_NO_DCI?
AFAIR we should release dci if no outstanding ops

Reply:
nop: in ```uct_dc_mlx5_iface_dci_put``` dci is released only in case if ep has no outstanding **and** iface **has** outstanding
@brminich am I right or missing something?

Reply:
discussed with @brminich - ```uct_dc_mlx5_iface_dci_has_outstanding(iface, ep->dci)``` is not needed, removed